Dart/src/com/jetbrains/lang/dart/ide/completion/DartServerCompletionContributor.java

```diff
@@ -99,93 +99,114 @@ protected void addCompletions(@NotNull final CompletionParameters parameters,
-               final String completionId = das.completion_getSuggestions(file, offset);
-               if (completionId == null) return;
-
-               final VirtualFile targetFile = file;
-               das.addCompletions(file, completionId, (replacementOffset, replacementLength, suggestion) -> {
-                 final CompletionResultSet updatedResultSet;
-                 if (uriPrefix != null) {
-                   updatedResultSet = resultSet;
-                 }
-                 else {
-                   final String specialPrefix = getPrefixForSpecialCases(parameters, replacementOffset);
-                   if (specialPrefix != null) {
-                     updatedResultSet = resultSet.withPrefixMatcher(specialPrefix);
-                   }
-                   else {
-                     updatedResultSet = resultSet;
-                   }
+
+               if (!das.getServerVersion().isEmpty() && StringUtil.compareVersionNumbers(das.getServerVersion(), "1.33") >= 0) {
+                 final DartAnalysisServerService.CompletionInfo2 completionInfo2 = das.completion_getSuggestions2(file, offset, 100);
+                 if (completionInfo2 == null || completionInfo2.mySuggestions.isEmpty()) {
+                   return;
                  }
 
-                 LookupElementBuilder lookupElement = null;
+                 final CompletionResultSet updatedResultSet = resultSet;
+                 List<CompletionSuggestion> suggestions = completionInfo2.mySuggestions;
 
-                 for (DartCompletionExtension extension : DartCompletionExtension.getExtensions()) {
-                   lookupElement = extension.createLookupElement(project, suggestion);
-                   if (lookupElement != null) break;
+                 // Add all the completion results that came back from the completion_getSuggestions2 call to this result set reference.
+                 for (CompletionSuggestion suggestion : suggestions) {
+                   updatedResultSet.addElement(createLookupElement(project, suggestion));
                  }
 
-                 if (lookupElement == null) {
-                   lookupElement = createLookupElement(project, suggestion);
+                 // As the user types additional characters, restart the completion only if we don't already have the complete set of
+                 // completions.
+                 if (completionInfo2.myIsIncomplete) {
+                   updatedResultSet.restartCompletionOnAnyPrefixChange();
                  }
 
-                 updatedResultSet.addElement(lookupElement);
-               }, (includedSet, includedKinds, includedRelevanceTags, libraryFilePathSD) -> {
-                 if (includedKinds.isEmpty()) {
-                   return;
+                 for (DartCompletionTimerExtension extension : DartCompletionTimerExtension.getExtensions()) {
+                   extension.dartCompletionEnd();
```

Comment:
If you don't mind, I'll move this right to the DartAnalysisServerService#completion_getSuggestions2() method, closer to `extension.dartCompletionStart()`. This will fix a possible code execution path when dartCompletionStart() is called but dartCompletionEnd() is not (notice `return` statement few lines above). It's easier to maintain code when both dartCompletionStart() and dartCompletionEnd() are in the same method.

Well, this way we'll not include lookup element creation in the gathered metrics but I guess it's <1% of the measured time.